sdk/core/Azure.Core.Experimental/src/PopTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,195 @@
+        /// Executed in the event a 401 response with a WWW-Authenticate authentication challenge header is received after the initial request.
+        /// </summary>
+        /// <remarks>Service client libraries may override this to handle service specific authentication challenges.</remarks>
+        /// <param name="message">The <see cref="HttpMessage"/> to be authenticated.</param>
+        /// <returns>A boolean indicating whether the request was successfully authenticated and should be sent to the transport.</returns>
+        protected virtual async ValueTask<bool> AuthorizeRequestOnChallengeAsync(HttpMessage message) =>
+            await AuthorizeRequestOnChallengeAsyncInternal(message, true).ConfigureAwait(false);
+
+        /// <summary>
+        /// Executed in the event a 401 response with a WWW-Authenticate authentication challenge header is received after the initial request.
+        /// </summary>
+        /// <remarks>Service client libraries may override this to handle service specific authentication challenges.</remarks>
+        /// <param name="message">The <see cref="HttpMessage"/> to be authenticated.</param>
+        /// <returns>A boolean indicating whether the request was successfully authenticated and should be sent to the transport.</returns>
+        protected virtual bool AuthorizeRequestOnChallenge(HttpMessage message) =>
+            AuthorizeRequestOnChallengeAsyncInternal(message, false).EnsureCompleted();
+
+        private async ValueTask AuthenticateAndAuthorizeRequestInternal(HttpMessage message, PopTokenRequestContext context, bool async)
+        {
+            _popNonce = context.ProofOfPossessionNonce;
+            var token = async ?
+                await _tokenCredential.GetTokenAsync(context, message.CancellationToken).ConfigureAwait(false) :
+                _tokenCredential.GetToken(context, message.CancellationToken);
+            message.Request.Headers.SetValue(HttpHeader.Names.Authorization, "PoP " + token.Token);
+        }
+
+        private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            if (message.Request.Uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS-protected (HTTPS) endpoints.");
+            }
+
+            if (async)
+            {
+                await AuthorizeRequestAsync(message).ConfigureAwait(false);
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                AuthorizeRequest(message);
+                ProcessNext(message, pipeline);
+            }
+
+            // Check if we have received a challenge or we have not yet issued the first request.
+            if (message.Response.Status == (int)HttpStatusCode.Unauthorized && message.Response.Headers.Contains(HttpHeader.Names.WwwAuthenticate))
```

Comment:
If this is another form of challenge authentication, do we also need to store and remove the body above before processing the next policy if we know we're going to fail with a 401?

Reply:
I think we do this for KeyVault for a specific reason related to discovering the scope. In this case, I don't know if there is any risk in keeping the body. Let me know if you feel strongly otherwise.

Reply:
My understanding is that it's primarily to avoid leaking secrets to third-parties when we have no Authentication header - to check that we actually get a valid one back first. But if you don't feel it necessary now...we can revisit later. I would check with MSAL what they think, though.